core/src/main/java/org/infinispan/interceptors/distribution/BaseDistributionInterceptor.java

```diff
@@ -144,288 +141,181 @@ public final BasicInvocationStage visitClearCommand(InvocationContext ctx, Clear
       return invokeNext(ctx, command);
    }
 
-   protected final CompletableFuture<InternalCacheEntry> retrieveFromProperSource(Object key, InvocationContext ctx,
-         FlagAffectedCommand command, boolean isWrite) throws Exception {
-      return doRetrieveFromProperSource(key, null, -1, ctx, command, isWrite);
-   }
-
-   private CompletableFuture<InternalCacheEntry> doRetrieveFromProperSource(Object key, InternalCacheEntry value,
-         int lastTopologyId, InvocationContext ctx, FlagAffectedCommand command, boolean isWrite) {
-      if (value != null)
-         return CompletableFuture.completedFuture(value);
-
-      final CacheTopology cacheTopology = stateTransferManager.getCacheTopology();
-      final int currentTopologyId = cacheTopology.getTopologyId();
+   protected final CompletableFuture<Void> remoteGet(InvocationContext ctx, AbstractDataCommand command,
+                                                     boolean isWrite) throws Exception {
+      final CacheTopology cacheTopology = checkTopologyId(command);
```

Comment:
> But I'd argue you need to learn how rebalancing works anyway in order to test all the possible cases.
You need to know how a rebalance happens, but you don't want to make assumptions which write owners will become read owners. We have to isolate concepts as much as possible.

> You mean the owner gained some keys in TS that were not in TR, right?
Yes, sorry. TR has less keys than TS but the time flows the other way.

> receiver to return an entry for which it is a write-owner, knowing that the new topology in which it is a read-owner was already sent out by the coordinator
Okay, I see that you can deduce that it already got all entries and confirmed it with current state transfer, but this is a complex and fragile assumption based on the state transfer implementation, not on actual consistent hashes. That's why I wanted to block there for the moment until the new topology comes and respond only afterward, as it does not know for sure (= from CH) that it can do that. So, if we send OTR (or rather UnsuccessfulResponse as the sender topology is not outdated), the sender has to retry.